### PR TITLE
Cce analytic data manager

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Adds the object for writing boundary data into the databox or outputting the analytic news to file.
The first two commits are to rework the linearized Bondi-Sachs solution to a different coordinate system, as it is better for working with the lapse and shift. I've also factored the checks of the ADM quantities so that the linearized Bondi-Sachs solution also now checks that those are well behaved.
Those changes are related because I use the linearized Bondi-Sachs solution as a test for checking that the analytic data manager is doing its job right.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [x] #2287

### CODE REVIEW

**Comment by moxcodes** [comment]
Rebased on develop, added dox and test for the new tag.

File: src/Evolution/Systems/Cce/AnalyticBoundaryData.hpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
+// Distributed under the MIT License.
```
Comment (moxcodes) [inline src/Evolution/Systems/Cce/AnalyticBoundaryData.hpp]: git mishap - this file should be removed.

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp [hunk 2]
```diff
@@ -54,6 +56,120 @@ extract_dr_bondi_scalars_from_cartesian_metric(
     const CartesianiSphericalJ& inverse_jacobian,
     const CartesianiSphericalJ& dr_inverse_jacobian,
     double extraction_radius) noexcept;
+
+// This function checks the consistency of the 3+1 ADM quantities in the
+// `boundary_tuple` with quantities computed from `expected_spacetime_metric`,
+// `expected_dt_spacetime_metric`, and `expected_d_spacetime_metric`. If the
+// expected quantities are also extracted from the tuple, this simply checks
+// that the boundary computation has produced a consistent set of quantities
+// and has not generated NaNs or other pathological values in the process.
```
Comment (kidder) [inline tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp]: what are other pathological values?

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp [hunk 3]
```diff
@@ -54,6 +56,120 @@ extract_dr_bondi_scalars_from_cartesian_metric(
...
+
+  const auto& pi =
+      get<GeneralizedHarmonic::Tags::Pi<3, ::Frame::Inertial>>(boundary_tuple);
+  const auto dt_spacetime_metric_from_pi =
+      GeneralizedHarmonic::time_derivative_of_spacetime_metric(
+          expected_lapse, expected_shift, pi, expected_d_spacetime_metric);
+  CHECK_ITERABLE_APPROX(expected_dt_spacetime_metric,
+                        dt_spacetime_metric_from_pi);
+
+  const auto& dt_lapse =
```
Comment (kidder) [inline tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp:116]: why do we need this block, how can it not be true if the pi is correct?

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.cpp [hunk 4]
```diff
@@ -42,6 +42,13 @@ extract_bondi_scalars_from_cartesian_metric(
       }
     }
   }
+  get<0, 0>(inverse_spherical_metric) =
```
Comment (kidder) [inline tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.cpp]: what are these changes for?  Should they be in another commit?

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.cpp [hunk 5]
```diff
@@ -123,6 +130,19 @@ extract_dt_bondi_scalars_from_cartesian_metric(
...
```
Comment (kidder) [inline tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.cpp]: just use `+=` or `-=` for these terms (also above and below)

File: tests/Unit/Evolution/Systems/Cce/Test_AnalyticBoundaryDataManager.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  UniformCustomDistribution<size_t> sdist{7, 10};
+  const size_t l_max = sdist(gen);
+  UniformCustomDistribution<double> parameter_dist{0.1, 1.0};
+  const double extraction_radius = 100.0 * parameter_dist(gen);
+  const double frequency = 0.01 * parameter_dist(gen);
+
+  const double time = 2.0;
+
+  Solutions::LinearizedBondiSachs analytic_solution{
+      {parameter_dist(gen) * 0.0, parameter_dist(gen) * 0.0},
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Test_AnalyticBoundaryDataManager.cpp]: huh?

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe

File: tests/Unit/Evolution/Systems/Cce/Test_AnalyticBoundaryDataManager.cpp (hunk 6 above)
Comment (moxcodes) [reply tests/Unit/Evolution/Systems/Cce/Test_AnalyticBoundaryDataManager.cpp]: hmm... suspicious. Must be a debugging remnant ... I'll revert.

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp (hunk 3 above)
Comment (moxcodes) [reply tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp:116]: These quantities are also retrieved from the `boundary_tuple`, so while it seems pretty redundant for the current implementation, it's not necessarily clear that the boundary calculation needs to compute these dt values from the `pi`, so it seemed best to include them in the general utility.

Comment (nilsdeppe) [reply tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp:116]: Sounds good! Maybe add a comment just so that they don't get deleted in the future?

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.cpp (hunk 4 above)
Comment (moxcodes) [reply tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.cpp]: These fix a mistake in confusing Boyer-lindquist time with retarded time that were caught by the ADM check. I think you're right that they should be in a separate commit.

**Comment by moxcodes** [comment]
fixups posted.
Cheers!

**Review by nilsdeppe — COMMENTED** [review]
LGTM! Once @kidder has had a chance to look, please go ahead and squash

**Comment by kidder** [comment]
you can squash

**Comment by moxcodes** [comment]
squashed.
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
